### PR TITLE
distcp should emit origin and upstream timestamp for dataset and file publish events

### PR DESCRIPTION
**Pull request by abti**
distcp should emit origin and upstream timestamp for dataset and file publish events

### CODE REVIEW

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableFile.java [hunk 1]
```diff
@@ -103,10 +109,12 @@
    * @param origin {@link FileStatus} of the original file.
    * @param datasetRoot Value of {@link CopyableDataset#datasetRoot} of the dataset creating this {@link CopyableFile}.
    * @param copyConfiguration {@link CopyConfiguration} for the copy job.
+   * @param originTimestamp Timestamp of file at its origin source.
+   * @param upstreamTimestamp Timestamp of file as in upstream.
    * @return a {@link CopyableFile.Builder}.
    */
   public static Builder builder(FileSystem originFs, FileStatus origin, Path datasetRoot,
-      CopyConfiguration copyConfiguration) {
+      CopyConfiguration copyConfiguration, long originTimestamp, long upstreamTimestamp) {
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableFile.java]: it should not be necessary to have the timestamps as arguments, that is what the builder is for.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/RecursiveCopyableDataset.java [hunk 2]
```diff
@@ -70,8 +70,10 @@ public OwnerAndPermission load(Path path) throws Exception {
     List<CopyableFile> copyableFiles = Lists.newArrayList();
 
     for (FileStatus file : files) {
-      copyableFiles.add(CopyableFile.builder(this.fs, file, this.rootPath, configuration).build());
+      copyableFiles.add(CopyableFile.builder(this.fs, file, this.rootPath, configuration, file.getModificationTime(),
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/RecursiveCopyableDataset.java]: these modifications are unnecessary if you don't change the builder signature.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java [hunk 3]
```diff
@@ -148,14 +149,26 @@ private void publishFileSet(CopyableFile.DatasetAndPartition datasetAndPartition
 
     fs.delete(datasetWriterOutputPath, true);
 
+    long datasetOriginTimestamp = Long.MAX_VALUE, datasetUpstreamTimestamp = Long.MAX_VALUE;
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: can you use separate lines per variable?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java [hunk 4]
```diff
@@ -119,7 +119,8 @@
                 CopyEventSubmitterHelper.DATASET_ROOT_METADATA_NAME, copyableDataset.datasetRoot().toString()));
             workUnit.setProp(SlaEventKeys.DATASET_URN_KEY, copyableDataset.datasetRoot().toString());
             workUnit.setProp(SlaEventKeys.PARTITION_KEY, copyableFile.getFileSet());
-            workUnit.setProp(SlaEventKeys.ORIGIN_TS_IN_MILLI_SECS_KEY, copyableFile.getFileStatus().getModificationTime());
+            workUnit.setProp(SlaEventKeys.ORIGIN_TS_IN_MILLI_SECS_KEY, copyableFile.getOriginTimestamp());
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java]: you don't need to set these properties anymore, the `CopyableFile` is available in the `WorkUnit`, and contains all this information.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java [hunk 5]
```diff
@@ -148,14 +149,26 @@ private void publishFileSet(CopyableFile.DatasetAndPartition datasetAndPartition
...
     for (WorkUnitState wus : datasetWorkUnitStates) {
       if (wus.getWorkingState() == WorkingState.SUCCESSFUL) {
         wus.setWorkingState(WorkUnitState.WorkingState.COMMITTED);
         CopyEventSubmitterHelper.submitSuccessfulFilePublish(eventSubmitter, wus);
       }
+      String originTimestampStr = wus.getProp(SlaEventKeys.ORIGIN_TS_IN_MILLI_SECS_KEY);
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: Better to deserialize the `CopyableFile` and get this information from there.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java [hunk 6]
```diff
@@ -148,14 +149,26 @@ private void publishFileSet(CopyableFile.DatasetAndPartition datasetAndPartition
...
+      String upstreamTimestampStr = wus.getProp(SlaEventKeys.UPSTREAM_TS_IN_MILLI_SECS_KEY);
+      long upstreamTimestamp = (null == upstreamTimestampStr) ? 0: Long.valueOf(upstreamTimestampStr);
+      if (upstreamTimestamp < datasetUpstreamTimestamp) {
+        datasetUpstreamTimestamp = upstreamTimestamp;
+      }
     }
 
-    CopyEventSubmitterHelper.submitSuccessfulDatasetPublish(eventSubmitter, datasetAndPartition);
+    CopyEventSubmitterHelper.submitSuccessfulDatasetPublish(eventSubmitter, datasetAndPartition,
+        datasetOriginTimestamp + "", datasetUpstreamTimestamp + "");
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: why not just call `toString`?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyEventSubmitterHelper.java [hunk 7]
```diff
@@ -38,10 +38,12 @@
   public static final String SIZE_IN_BYTES = "SizeInBytes";
 
   static void submitSuccessfulDatasetPublish(EventSubmitter eventSubmitter, CopyableFile.DatasetAndPartition
-      datasetAndPartition) {
+      datasetAndPartition, String originTimestamp, String upstreamTimestamp) {
     SlaEventSubmitter.builder().eventSubmitter(eventSubmitter).eventName(DATASET_PUBLISHED_EVENT_NAME)
         .datasetUrn(datasetAndPartition.getDataset().getDatasetRoot().toString())
         .partition(datasetAndPartition.getPartition())
+        .originTimestamp(originTimestamp)
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyEventSubmitterHelper.java]: You should also add similar changes to `submitSuccessfulFilePublish`.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableFile.java (hunk 1 above)
Comment (abti) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableFile.java]: Changed

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/RecursiveCopyableDataset.java (hunk 2 above)
Comment (abti) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/RecursiveCopyableDataset.java]: Yeah, got removed

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java (hunk 3 above)
Comment (abti) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: Moved

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java (hunk 4 above)
Comment (abti) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java]: Yeah, got removed

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java (hunk 5 above)
Comment (abti) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: Changed

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java (hunk 6 above)
Comment (abti) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: Changed

File: gobblin-data-management/src/test/java/gobblin/data/management/copy/CopyableFileTest.java [hunk 8]
```diff
@@ -33,7 +33,7 @@ public void testSerializeDeserialze() throws Exception {
         new CopyableFile(new FileStatus(10, false, 12, 100, 12345, new Path("/path")), new Path("/destination"),
             new Path("/relative"), new OwnerAndPermission("owner", "group", FsPermission.getDefault()),
             Lists.newArrayList(new OwnerAndPermission("owner2", "group2", FsPermission.getDefault())),
-            "checksum".getBytes(), PreserveAttributes.fromMnemonicString(""), "");
+            "checksum".getBytes(), PreserveAttributes.fromMnemonicString(""), "", 0, 0);
```
Comment (ibuenros) [inline gobblin-data-management/src/test/java/gobblin/data/management/copy/CopyableFileTest.java]: This signature is wrong isn't it? Last two arguments shouldn't be there. Similar for other test changes.

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java [hunk 9]
```diff
@@ -148,14 +147,25 @@ private void publishFileSet(CopyableFile.DatasetAndPartition datasetAndPartition
...
 
+    long datasetOriginTimestamp = Long.MAX_VALUE;
+    long datasetUpstreamTimestamp = Long.MAX_VALUE;
+
     for (WorkUnitState wus : datasetWorkUnitStates) {
       if (wus.getWorkingState() == WorkingState.SUCCESSFUL) {
         wus.setWorkingState(WorkUnitState.WorkingState.COMMITTED);
         CopyEventSubmitterHelper.submitSuccessfulFilePublish(eventSubmitter, wus);
       }
+      CopyableFile copyableFile = CopySource.deserializeCopyableFiles(wus).get(0);
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyDataPublisher.java]: Can you test this in the `CopyableFile` test?

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableFile.java [hunk 10]
```diff
@@ -186,9 +191,15 @@ public CopyableFile build() throws IOException {
       if (this.fileSet == null) {
         this.fileSet = this.rootPath.toString();
       }
+      if (this.originTimestamp == 0) {
```
Comment (ibuenros) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableFile.java]: Can you test this in the CopyableFile test?

File: gobblin-data-management/src/test/java/gobblin/data/management/copy/CopyableFileTest.java (hunk 8 above)
Comment (abti) [inline gobblin-data-management/src/test/java/gobblin/data/management/copy/CopyableFileTest.java]: As discussed, continued using the constructor but added tests for builder

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableFile.java (hunk 10 above)
Comment (abti) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/CopyableFile.java]: Tested with builder test

File: gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyEventSubmitterHelper.java (hunk 7 above)
Comment (abti) [inline gobblin-data-management/src/main/java/gobblin/data/management/copy/publisher/CopyEventSubmitterHelper.java]: submitSuccessfulFilePublish is getting these from copyablefile in WUS

**Comment by ibuenros** [comment]
LGTM

**Pull request merged**